### ISSUE TITLE
KFServing docs are impossible to follow

### ISSUE THREAD

**Issue by ulrikpl** [kubeflow/website#2589]
I am walking through the README.md as it appears to me with all the inconsistencies, confusion and dead ends that meets me as a KFServing newbie failing to set it up for even a demo project. Forgive me for playing devils advocat, I know it is open source and people do it for free, but these are frustrations and confusions that I think many users meet when failing to get this to work. So if nothing else perhaps you can use it to make docs more user friendly at some point.

NB: Installed KF1.2 on GCP as per Kubeflow guide

```
Docs:
cluster-local-gateway is required to serve cluster-internal traffic for transformer and explainer use cases. 
Please follow instructions here to install cluster local gateway.
```

The Kubeflow 1.2 GCP install guide installed all kinds of stuff so I have no idea if I already have such a "cluster local gateway" and have no idea on how to look for it. Would be very helpful with a hint or line of code to find out if it's installed already or not. 

I guess I better try to install it. Following the link to 'install cluster local gateway' the first lines say:

```
Before you begin
You need:
A Kubernetes cluster created.
istioctl (v1.7 or later) installed.
```

But I have istioctl client version: 1.4.10-asm.18 as required by the Kubeflow 1.2 GCP install guide (special Anthos Service Mesh version). So that didn't work and I guess I will have to live without "transformer" and "explainer" functionality in my KFServing (unless I have a cluster-local-gateway already, which I don't know how to check for).

Next up is this:
```
If you are looking to use PodSpec fields such as nodeSelector, affinity or tolerations which are now supported in the KFServing
v1beta1 API spec, this requires Knative v0.17.0+, and you need to turn on the corresponding feature flags in your Knative 
configuration.
```

Again having followed the latest Kubeflow 1.2 GCP install guide I have API version v1alpha2 and Knative v0.14.3. So nodeSelectors I guess are out too and I wonder if that means that big models can't be served on their own node pools and whether I therefore will have to upgrade my Kubeflow cluster node pool to an expensive beast running 24/7. There are no info about how to upgrade either Knative or API version - not to even a mention if it is at all possible.

Next comes "Setup Ingress Gateway & Determine the ingress IP and ports" and I soon learn that I have no external load balancer to use for the ingress gateway.

```
$ kubectl get svc istio-ingressgateway -n istio-system
NAME                   TYPE       CLUSTER-IP    EXTERNAL-IP   PORT(S)     AGE 
istio-ingressgateway   NodePort   10.7.247.24   <none>        15020:30681/TCP,...,15443:32690/TCP   15d
```

I read:
```
Docs:
If the EXTERNAL-IP value is none (or perpetually pending), your environment does not provide an external load balancer
for the ingress gateway. In this case, you can access the gateway using the service’s node port.
```

There is no information what the consequences of this are. Won't I be able to load balance my ML prediction service without an "external IP"? Is that just it? Is there no way to add such an external IP or would it screw up the whole system? The lack of nodeSelector I guess means that I can't scale to zero since I can't use a dedicated node pool, but does this lack of external IP mean that I can't auto scale a prediction server at all?

Next step is this:
```
# GKE
export INGRESS_HOST=worker-node-address
```

Unfortunately there is no information about how to find this "worker-node-address"? Or is it just a string perhaps? There are no indications that it is a variable but no pods or nodes have a label called 'worker-node-address'. Perhaps it is the CLUSTER-IP from before or maybe the value for "Node" I get when describing the gateway pod?

```
$ kubectl describe pod istio-ingressgateway-xxxx -n istio-system
```
Yielding:
```
Name:         istio-ingressgateway-xxxxxxxxx-55wwv
Namespace:    istio-system
Priority:     0
Node:         gke-kubeflow-default-pool-xxxxxx-xxxx/10.132.0.11
... etc.
```

I have no clue so will have to skip that option too.

```
Docs:
Alternatively you can do Port Forward for testing purpose
```

Thank god, at least I can get to play with it locally until someone makes a youtube video explaining it to dummies like me.

I follow the step to 'port forward' and continues to "testing the KFServing" installation like described:

```
Docs:
kubectl get po -n kfserving-system
NAME                             READY   STATUS    RESTARTS   AGE
kfserving-controller-manager-0   2/2     Running   2          13m
```

Is that the same as 'kubectl get pods -n kfserving-system'? It doesn't really matter because there is no namespace called kfserving-system!

```
$kubectl get namespaces
NAME              STATUS   AGE
cert-manager      Active   15d
default           Active   15d
default-profile   Active   15d
istio-system      Active   15d
knative-serving   Active   15d
kube-node-lease   Active   15d
kube-public       Active   15d
kube-system       Active   15d
kubeflow          Active   15d
```

Nevermind, I will just cross my fingers and hope that it is working and continue to "Create KFServing test inference service" with this:

```
Docs:
API_VERSION=v1beta1
kubectl create namespace kfserving-test
kubectl apply -f docs/samples/${API_VERSION}/sklearn/v1/sklearn.yaml -n kfserving-test
```

It needs an 'export API_VERSION', but I could guess that since it didn't work.

and
```
kubectl get inferenceservices sklearn-iris -n kfserving-test
NAME           URL                                                 READY   PREV   LATEST   PREVROLLEDOUTREVISION   LATESTREADYREVISION                    AGE
sklearn-iris   http://sklearn-iris.kfserving-test.example.com      True           100                              sklearn-iris-predictor-default-47q2g   7d23h
```

Happy days demo is running. I haven't configured a DNS, so will skip "Curl with real DNS" and jump to "Curl with magic DNS":

```
Docs:
If you don't want to go through the trouble to get a real domain, you can instead use "magic" dns xip.io.
The key is to get the external IP for your KFServing cluster
```

Seriously? I wouldn't be down here in localhost land if there was an external IP for the KFServing cluster.

```
Docs
kubectl get svc istio-ingressgateway -n istio-system

Look for the EXTERNAL-IP column's value
```

Which like before gives this:
```
NAME                   TYPE       CLUSTER-IP    EXTERNAL-IP   PORT(S)     AGE 
istio-ingressgateway   NodePort   10.7.247.24   <none>        15020:30681/TCP,...,15443:32690/TCP   15d
```

And I need a pause to go bang my head against a wall...

**Comment by thesuperzapper** [comment]
@kubeflow/wg-serving-leads sounds like you might need to look at your docs!

**Comment by yuzisun** [comment]
@ulrikpl Thanks for the feedbacks! The main KFServing doc is more focusing on the standalone KFServing and does not assume you have Kubeflow. If you are using KFServing from Kubeflow install by default you shouldn't  need to install anything extra yourself, but the KFServing/Knative version in Kubeflow is generally a few versions behind and if you are looking to use the latest versions you would need to upgrade yourself or wait until they are integrated to Kubeflow. I think in Kubeflow 1.3 we are trying to make this process easier for user so you can take and upgrade individual component easily.

To answer some of your questions:
- Standalone KFServing installs in `kfserving-system` while Kubeflow installs KFServing in `kubeflow` namespace, again the KFServing doc is more from the angle of standalone KFServing.
- For load balancer vs node port, I would recommend reading [this article](https://medium.com/google-cloud/kubernetes-nodeport-vs-loadbalancer-vs-ingress-when-should-i-use-what-922f010849e0), I think your problem is more GKE specific and for demo purpose you could use node ip with host header. I with we had a better doc for GKE/GCP KFServing but [this](https://github.com/kubeflow/kfserving/tree/master/docs/samples/gcp-iap) is what we currently have, feel free to improve and add your stuff there. 

 I am happy to work with you to improve our docs! thanks again for the feedbacks!

**Comment by ulrikpl** [comment]
@yuzisun Thanks for the reply. I think you should at least include a disclaimer or clear warning in the top of the KFS readme, so that the next user who comes along doesn’t waste precious time trying to get these things to work on their newly installed GKE kubeflow cluster. 

It might seem glaringly obvious to the developers of Kubeflow but I would guess that most users simply wish to install and use Kubeflow or KFServing and don’t have years of experience working with Istio, Knative or K8s. Such knowledge is not at all trivial to acquire so users need a lot more hand holding than what you might think. Trying to figure things out on your own without proper prerequisites is just one painful rabbit hole after the next without any progress.

I went ahead as you suggested and attempted to install a dedicated KFServing cluster on GKE and have failed miserably due to the utterly useless docs. They are so rudimentary and so filled with contradictions that I suspect only the Kubeflow development team itself can make a successful install. 

**Let me give a few of examples of what I run into trying to set this up:**

The first problem and major confusion is what combination of Istio, Knative and KFServing we should install (obviously we want KFServing v0.5+ or we wouldn’t be here)

KFServing docs (https://github.com/kubeflow/kfserving) state following prerequisites:

Istio: v1.9.0
Knative Serving: v.0.17.4+

notice that everything I read on the Knative page is for Knative v0.22.

Q: You write that Knative v0.17.4+ works but did anybody really try it with v0.22? In the developer docs you suggest v0.20.

Q:Are Knative’s Istio install guidelines the same for Knative v0.22 as they were for Knative v0.17.4? I don’t know because the drop down menu in Knative docs only let you go back to Knative v0.19 and no earlier archives or versions mention how to install Istio at all. If we choose Knative v0.20 in the Knative docs we are redirected to a site with no mention of Istio.

Q: So again, which versions should we really use? What is a latest safe working combination of Istio, Knative, KFServing that requires the least amount of changes and configurations downstream.

Q: Wouldn’t it be a LOT easier to just put a workable set of commands in the KFserving docs rather than linking to Knative who constantly updates their page making the KFServing docs out of sync and thus useless?

**Install Istio**

I decide to hold my nose and try to install Istio v1.7.1 as suggested in the developer page. First thing in the docs I run into is this:

```
If you want to get up running Knative quickly or you do not need service mesh, 
we recommend installing Istio without service mesh(sidecar injection).
```

Here is the problem: I am not on this page to ‘get up and running with Knative quickly’. I am here to use KFServing. As a first time user I obviously don’t know the inner workings of KFServing so how could I possibly know whether I need service mesh and sidecar injection or not? Why don’t you just tell users what to install and how?

As so often before with anything KF docs I resort to hope as a strategy and install it without sidecar injection in order to keep things simple.

Further down in the Istio install guide I reach Istio mTLS which is described as an optional production feature. I don’t need more complex I need less so decide to skip it for now.

Under the Istio mTLS paragraph I read this:

```
After you install the cluster local gateway, your service and deployment for the local gateway is named knative-local-gateway.
```

When did I install ‘cluster local gateway’? How do I know if I have it installed? What is it? What does it look like? What does it have to do with mTLS? What service are they talking about? What deployment? I skipped the mTLS as it was stated as optional and brought complexity, but does that mean I have no ‘cluster local gateway’? Back in the KFserving docs it says that I should install ‘cluster local gateway’ so it doesn’t seem optional at all. Will I automatically have ‘cluster local gateway’ installed if I enable mTLS. Why don’t you just put some working commands into the KFS docs instead of linking to forever changing third party docs.

Next up on my journey to install Istio:

```
Updating the config-istio configmap to use a non-default local gateway
If you create a custom service and deployment for local gateway with a name other than knative-local-gateway, 
you need to update gateway configmap config-istio under the knative-serving namespace.
```

How could I possibly know if am going to use a ‘non-default local gateway’? Does KFServing use a ‘non-default local gateway’? Should I skip it? You should know, so why don’t you just tell us?

Next up:

```1. Edit the config-istio configmap:```

I try the command and get this:
```
$ Kubectl get configmaps -n istio-system
NAME                                   DATA   AGE
istio                                  2      98m
istio-ca-root-cert                     1      98m
istio-leader                           0      98m
istio-namespace-controller-election    0      98m
istio-sidecar-injector                 2      98m
istio-validation-controller-election   0      98m
kube-root-ca.crt                       1      98m
```

**There are no configmaps named ‘config-istio’!**

It seems that it might belong to Knative Serving, but we haven’t even installed that yet as per KFserving docs install order. I guess I should first try to install Knative Serving and come back to this.


### Knative Serving Install:

The developer docs said that v0.20 had been tested with Istio 1.7.1 that I just installed, so I will go with that version and see what happens.

I edit the serving-core.yaml to enable nodeSelectors and hope that I did it correctly.

```
Since Knative v0.19.0 cluster local gateway deployment has been removed and shared with ingress gateway, 
if you are on Knative version later than v0.19.0 you should modify localGateway to knative-local-gateway and 
localGatewayService to knative-local-gateway.istio-system.svc.cluster.local in the inference service config.
```

I guess don’t have to worry about this ‘cluster local gateway’ anyways. I just need to modify the inference
 service config. But I have no inference service running! We haven’t even installed KFServing yet, so I assume that 
 this fix is simply placed in in wrong order and I will have to get back to it later.

**Installing Certs:**
I try use your "self-assigned certs generation" script but it fails:
 ```
 Error from server (NotFound): error when creating "STDIN": namespaces "kfserving-system" not found
 ```

I guess it is not an install afterall but something to run after I installed KFServing. I cross my fingers and decide to try it out after installing KFServing.

**Install KFServing:**

I decide to use TAG=v0.5.0 and not v0.5.1 based on the Istio and Knative Serving versions I have installed and the fact that those had been tested together successfully according to the developer docs. 

```
kubectl apply -f https://github.com/kubeflow/kfserving/releases/download/v0.5.0/kfserving.yaml

…
…
…
unable to recognize "https://github.com/kubeflow/kfserving/releases/download/v0.5.0/kfserving.yaml": no matches for kind 
"Certificate" in version "cert-manager.io/v1alpha2"
unable to recognize "https://github.com/kubeflow/kfserving/releases/download/v0.5.0/kfserving.yaml": no matches for kind
 "Issuer" in version "cert-manager.io/v1alpha2"
make: *** [install_kfserving] Error 1
```

So that didn’t work. It seems I could not just use the “self-assigned certs generation script” after installing KFServing and I actually has to install Certs anyways. Seems that the script wasn’t much of an alternative after all.

Curiosity got the better of me so after installing Certs so had to try running the “self-assigned certs generation script” after installing Certs. Wow! All sorts of stuff happened with keys and stuff flying around. I wonder if I broke or saved my install. It is clear that these two methods are not alternatives. The script doesn’t work without certs and certs without the script would be in a very different state.

I continue and ...

**install KFServing v0.5.0**

And … it doesn’t work.

```
kfserving-system   kfserving-controller-manager-0                              1/2     CrashLoopBackOff   8          21m
```

Wonder why this is, but it probably has something to do with that certs script as that is the primary difference to my previous 3 unsuccessful attempts at getting a workable KFServing install.

I delete KFServing v0.5.0 and try v0.5.1… which surprisingly seemed to do the trick! All containers are running.

I create the sklearn test inference service as per the docs and test it like this:

```
kubectl get inferenceservices sklearn-iris -n kfserving-test

NAME           URL   READY     PREV   LATEST   PREVROLLEDOUTREVISION   LATESTREADYREVISION                    AGE
sklearn-iris         Unknown          100                              sklearn-iris-predictor-default-00002   4m11s
```

Yet another fail. 

I go to the debug guide and first time around trying the commands I get all empty returns. Turns out that the docs have missed including namespaces, which I then proceed to do.

```
kubectl get ksvc -n kfserving-test

```

And I learn this: IngressNotConfigured

Next up I check the pod logs as per guide:

```
You can check Knative networking-istio pod logs for more details.
$ kubectl logs -l app=networking-istio -n knative-serving
>
$
```

and get nothing returned. Turns out there are no resources labelled ‘networking-istio’

```
$kubectl get pods -l app=networking-istio -A
No resources found
```

I don’t know where to look for a solution but remember that I skipped ‘Updating the config-istio configmap to use a non-default local gateway’ when installing Istio. Nothing in there gave any sense for someone who didn’t build KFserving but I have no better ideas.

I can however see there there is still no configmap called "config-istio". Not that I knew what it did anyways.

```
$Kubectl get configmaps -n istio-system
NAME                                   DATA   AGE
istio                                  2      4h6m
istio-ca-root-cert                     1      4h6m
istio-leader                           0      4h6m
istio-namespace-controller-election    0      4h6m
istio-sidecar-injector                 2      4h6m
istio-validation-controller-election   0      4h6m
kube-root-ca.crt                       1      4h6m
```

Notice how the AGE has now increased from 98m to 4h6m. That is 3 and half additional hours wasted. It seems to me that this istio-minimal-operator.yaml that you recommend is a bit too minimal.

And after two full days of failing to get this working I am once again looking for a wall to bang my poor old head against.

It is paradoxical that KFServing aimed to abstract away complex configurations making life easier and speed to production faster. Yet the docs are so rudimentary and constructed using forever changing outside links that just installing the basic tool appears impossible for all but the people who built it. Even the demos I have tried are broken. For instance the Bert demo relies on a github repo that have since seen numerous commits and no longer exists in the form that the demo was written for. Files are missing and directory structures changed so everything comes to a squeaking halt at the very first command. It is symptomatic and reminds me of a music band going through the trouble of writing songs and recording great albums and then proceed to stuff away all the records in the attic so that no one can enjoy their hard work.

**Comment by thesuperzapper** [comment]
@yuzisun @ulrikpl did anything come from this, or can it be closed?

**Comment by ulrikpl** [comment]
Nothing came of it but since V1.3 is out it is no longer relevant. That said V1.3 install is much easier but KFServing docs are still not very useful as it is near impossible to get a simple model server to work on GKE.

**Comment by asok-mirror** [comment]
I have a doubt here,

looks like the deployment service doest have the node port assigned to access the app outside the cluster in local env. 

Is the only way to test the service through port forwarding?

I tried with the manifest supplied in sklearn-iris sample v2.

Is there a way to have the node port assigned to the pod service?

Beginner here, pls apologize for any

@ulrikpl please let me know if you can help me here.

**Comment by ulrikpl** [comment]
I am sorry asok-mirror but my company eventually gave up on KFServing due to the lack of documentation and the spaghetti factory of technologies like Istio and Knative that without proper documentation made it too too time consuming/expensive to debug. The demos given are rudimentary (nothing on transformers for example) and simply didn't work in our GCP environment. IPs of model services came and went with upgrades for no obvious reasons. We wanted to use load balancer so nodeports were not an option. In essence we found it insufficiently mature to run crucial services for our clients on so have looked elsewhere.
